packages/react/popper/src/Popper.tsx

```diff
@@ -105,71 +126,136 @@ const PopperContent = React.forwardRef<PopperContentElement, PopperContentProps>
+    const arrowWidth = arrowSize?.width ?? 0;
+    const arrowHeight = arrowSize?.height ?? 0;
+
+    const desiredPlacement = (side + (align !== 'center' ? '-' + align : '')) as Placement;
+
+    const { reference, floating, strategy, x, y, placement, middlewareData, update } = useFloating({
+      // default to `fixed` strategy so users don't have to pick and we also avoid focus scroll issues
+      strategy: 'fixed',
+      placement: desiredPlacement,
+      whileElementsMounted: autoUpdate,
+      middleware: [
+        offset({ mainAxis: sideOffset + arrowHeight, alignmentAxis: alignOffset }),
+        avoidCollisions
+          ? shift({
+              mainAxis: true,
+              crossAxis: false,
+              padding: collisionTolerance,
+              limiter: limitShift(),
+            })
+          : undefined,
+        arrow ? floatingUIarrow({ element: arrow, padding: arrowPadding }) : undefined,
+        avoidCollisions ? flip({ padding: collisionTolerance }) : undefined,
+        transformOrigin({ arrowWidth, arrowHeight }),
+      ].filter(isDefined),
+    });
 
-    const composedRefs = useComposedRefs(forwardedRef, (node) => setContent(node));
+    // assign the reference dynamically once `Content` has mounted so we can collocate the logic
+    useLayoutEffect(() => {
+      reference(context.anchor);
+    }, [reference, context.anchor]);
 
-    const windowSize = useWindowSize();
-    const collisionBoundariesRect = windowSize
-      ? DOMRect.fromRect({ ...windowSize, x: 0, y: 0 })
-      : undefined;
-
-    const { popperStyles, arrowStyles, placedSide, placedAlign } = getPlacementData({
-      anchorRect,
-      popperSize: contentSize,
-      arrowSize,
-
-      // config
-      arrowOffset,
-      side,
-      sideOffset,
-      align,
-      alignOffset,
-      shouldAvoidCollisions: avoidCollisions,
-      collisionBoundariesRect,
-      collisionTolerance,
-    });
-    const isPlaced = placedSide !== undefined;
+    const isPlaced = x !== null && y !== null;
+    const [placedSide, placedAlign] = getSideAndAlignFromPlacement(placement);
+    const { arrow: { x: arrowX, y: arrowY } = {} } = middlewareData;
+    const cannotCenterArrow = middlewareData.arrow?.centerOffset !== 0;
 
     const [contentZIndex, setContentZIndex] = React.useState<string>();
     useLayoutEffect(() => {
       if (content) setContentZIndex(window.getComputedStyle(content).zIndex);
     }, [content]);
 
+    const { hasParent, positionUpdateFns } = usePositionContext(CONTENT_NAME, __scopePopper);
+    const isRoot = !hasParent;
+
+    React.useLayoutEffect(() => {
+      if (!isRoot) {
+        positionUpdateFns.add(update);
+        return () => {
+          positionUpdateFns.delete(update);
+        };
+      }
+    }, [isRoot, positionUpdateFns, update]);
+
+    // when nested contents are rendered in portals, they are appended out of order
+    // we need to re-compute the positioning once the parent has finally been placed.
```

Comment:
```suggestion
    // when nested contents are rendered in portals, they are appended out of order causing
    // children to be positioned incorrectly if initially open.
    // we need to re-compute the positioning once the parent has finally been placed.
    // https://github.com/floating-ui/floating-ui/issues/1531
```

How'd you feel about being really specific on this one?